### PR TITLE
RUM-3535 feat: add support for trace context injection configuration to allow selective injection

### PR DESCRIPTION
**Pull request by ganeshnj**
### What and why?

RFC https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/3529509193

### How?

There is one big difference has been in this PR along with the implementation of the RFC.  If we have a distributed span which has a parent, the parent's sampling strategy will be applied to the distributed span.

Now if the parent span is not sampled, the distributed span will not be sampled regardless of the distributed sampling configuration and vice versa.

This is similar to the Trace Agent behavior where the decision stays consistent across the board on the spans. This way customers will not have a broken trace ever.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [x] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [x] Run unit tests for Core, RUM, Trace, Logs, CR and WVT
- [ ] Run unit tests for Session Replay
- [x] Run integration tests
- [ ] Run smoke tests
- [ ] Run tests for `tools/`

### CODE REVIEW

**Review by maxep — DISMISSED** [review]
Looks great, very clear!

File: DatadogRUM/Sources/RUMConfiguration.swift [hunk 1]
```diff
@@ -294,13 +294,23 @@ extension RUM.Configuration.URLSessionTracking {
         /// - Parameters:
         ///   - hosts: The set of hosts to inject tracing headers. Note: Hosts must not include the "http(s)://" prefix.
         ///   - sampleRate: The sampling rate for tracing. Must be a value between `0.0` and `100.0`. Default: `20`.
-        case trace(hosts: Set<String>, sampleRate: Float = 20)
+        ///   - traceControlInjection: The strategy for injecting trace context into requests. Default: `.all`.
+        case trace(
+            hosts: Set<String>,
+            sampleRate: Float = 20,
+            traceControlInjection: TraceContextInjection = .all
```
Comment (ncreated) [inline DatadogRUM/Sources/RUMConfiguration.swift:301]: question/ I wonder if the `TraceContextInjection` type shouldn't be re-exported through `DatadogRUM` [like we do for other internal types](https://github.com/DataDog/dd-sdk-ios/blob/c044ef9c69369a44c772d395c838d8c6951c2ab3/DatadogRUM/Sources/RUMConfiguration.swift#L15). My feeling is we don't need it unless someone wants to define a property of this type in their code, which is not the case here. For extra check, we can call this API in [smoke tests](https://github.com/DataDog/dd-sdk-ios/blob/c044ef9c69369a44c772d395c838d8c6951c2ab3/dependency-manager-tests/spm/App/ViewController.swift#L20) to see if it works without imoprting `DatadogInternal`.

The same applies to [re-exporting types in `DatadogTrace`](https://github.com/DataDog/dd-sdk-ios/blob/c044ef9c69369a44c772d395c838d8c6951c2ab3/DatadogTrace/Sources/TraceConfiguration.swift#L11-L21).

File: DatadogInternal/Tests/NetworkInstrumentation/HTTPHeadersWriterTests.swift [hunk 2]
```diff
@@ -28,7 +28,7 @@ class HTTPHeadersWriterTests: XCTestCase {
     }
 
     func testWritingDroppedTraceContext_withAutoSamplingStrategy() {
```
Comment (ncreated) [inline DatadogInternal/Tests/NetworkInstrumentation/HTTPHeadersWriterTests.swift:30]: suggestion/ I think the `*AutoSamplingStrategy*` is a leftover after renaming `samplingStrategy` from `.auto` to `.headBased`. Could we update this name in all tests, so it is correct? 🏕️ 🧹

File: DatadogInternal/Tests/NetworkInstrumentation/W3CHTTPHeadersReaderTests.swift
Comment (ncreated) [inline DatadogInternal/Tests/NetworkInstrumentation/W3CHTTPHeadersReaderTests.swift:1]: change-request/ I only see updates to existing tests with `.all` injection, but no new tests added for `.sampled`. Aren't we missing enough coverage for new variants?

File: DatadogInternal/Tests/NetworkInstrumentation/W3CHTTPHeadersWriterTests.swift
Comment (ncreated) [inline DatadogInternal/Tests/NetworkInstrumentation/W3CHTTPHeadersWriterTests.swift:1]: change-request/ Similar to `W3C*ReaderTests`, shall we add coverage to new variants with using `.sampled` injection?

**Review by ncreated — CHANGES_REQUESTED** [review]
Looks good and reads really well 🎯 👍

My only concern is on added tests coverage - it seems we updated existing unit tests for Header Writers and Readers 👍, but we haven't added coverage for new permutations of their `init()`, no?

File: DatadogRUM/Sources/RUMConfiguration.swift (hunk 1 above)
Comment (ganeshnj) [reply DatadogRUM/Sources/RUMConfiguration.swift:301]: actually we should, else customer have to import internal. Good call.

File: DatadogInternal/Tests/NetworkInstrumentation/HTTPHeadersWriterTests.swift (hunk 2 above)
Comment (ganeshnj) [reply DatadogInternal/Tests/NetworkInstrumentation/HTTPHeadersWriterTests.swift:30]: sure

File: DatadogInternal/Tests/NetworkInstrumentation/W3CHTTPHeadersWriterTests.swift
Comment (ganeshnj) [reply DatadogInternal/Tests/NetworkInstrumentation/W3CHTTPHeadersWriterTests.swift:1]: Good call.

File: DatadogInternal/Tests/NetworkInstrumentation/W3CHTTPHeadersReaderTests.swift
Comment (ganeshnj) [reply DatadogInternal/Tests/NetworkInstrumentation/W3CHTTPHeadersReaderTests.swift:1]: Good call.

**Review by ncreated — DISMISSED** [review]
🎯

**Review by maciejburda — DISMISSED** [review]
LGTM

**Review by maxep — APPROVED** [review]

**Review by maxep — APPROVED** [review]

**Pull request merged**